Src/FluentAssertions/Collections/MaximumMatching/MaximumMatchingProblem.cs

```diff
@@ -0,0 +1,69 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using FluentAssertions.Collections.MaximumMatching;
+
+namespace FluentAssertions.Collections
+{
+    /// <summary>
+    /// The <see cref="MaximumMatchingProblem{TElement}"/> class defines input for the maximum matching problem.
+    /// The input is a list of predicates and a list of elements.
+    /// The goal of the problem is to find such mapping between predicates and elements that would maximize number of matches.
+    /// A predicate can be mapped with only one element.
+    /// An element can be mapped with only one predicate.
+    /// </summary>
+    /// <typeparam name="TElement">The type of elements which must be matched with predicates.</typeparam>
+    internal class MaximumMatchingProblem<TElement>
+    {
+        private readonly Dictionary<int, List<int>> matchingElementsByPredicate = new Dictionary<int, List<int>>();
```

Comment:
Thanks a lot for the valuable feedback, Dennis! I'm on it :)

I'm not 100% sure about the suggestion below:
_Since the namespace is already called MaximumMatching, I would repeat that word on every class and method._
Removing the MaximumMatching prefix from class names would leave us with classes called Problem, Solution and Solver. These names are very generic. It would be hard to understand what the calling code is doing. Examples from well-known libraries where prefix is used: SqlConnection instead of Connection, CosmosClient instead of Client. I think keeping the prefix would be the right thing to do. But pls let me know if you'd still rather rename.

Reply:
It was a suggestion anyway, so it's up to you. And since this code is `internal` anyway, we can always rename it afterwards if we change our minds again.

Reply:
I hope I have addressed the comments in the latest iteration. Also, I know that the matching problem is adding a lot of complexity to the project. Apologize for it. I tried making it as readable as possible and any suggestions to make it better are very welcome.